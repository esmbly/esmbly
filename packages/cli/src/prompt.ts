import inquirer from 'inquirer';
import { Config } from '@esmbly/types';
import {
  getAvailableOutputFormats,
  getAvailableTransformers,
} from '@esmbly/utils';

export async function promptForConfig(): Promise<Config> {
  const { input } = await inquirer.prompt([
    {
      message: 'Which files do you want to transform?',
      name: 'input',
      type: 'input',
    },
  ]);

  const { transformers } = await inquirer.prompt([
    {
      choices: await getAvailableTransformers(),
      message: 'Which transformer do you want to use?',
      name: 'transformers',
      type: 'checkbox',
    },
  ]);

  // TODO: Ask to install transformers here?

  const { output } = await inquirer.prompt([
    {
      choices: await getAvailableOutputFormats(transformers),
      message: 'Which output formats to you want to use?',
      name: 'output',
      type: 'checkbox',
    },
  ]);

  return { input: [input], output, transformers };
}
