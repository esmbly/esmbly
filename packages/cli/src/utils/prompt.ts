import inquirer from 'inquirer';
import {
  TransformerInfo,
  getAvailableOutputFormats,
  getAvailableTransformers,
} from '@esmbly/utils';
import stringify from 'stringify-object';
import chalk from 'chalk';

interface PromptedConfig {
  module: string;
  install: boolean;
  transformers: TransformerInfo[];
  pkgManager: string;
}

function createRequireString(transformers: TransformerInfo[]): string {
  return transformers
    .map((t: TransformerInfo) => `const ${t.const} = require('${t.pkg}')`)
    .join('\n');
}

function createConfigString(
  input: string,
  output: string[],
  outDir: string,
): string {
  return stringify({
    input: [input],
    transformers: '<TRANSFORMERS>',
    // eslint-disable-next-line
    output: output.map((format: string) => ({ format, outDir })),
  });
}

function createTransformerCalls(transformers: TransformerInfo[]): string {
  /**
   * Produces:
   * [
   *   Flow.createTransformer(),
   *   Wasm.createTransformer()
   * ]
   */
  return `${transformers.reduce(
    (calls, t) => `${calls}\n    ${t.const}.createTransformer(),`,
    '[',
  )}\n  ]`;
}

function createModuleString(
  input: string,
  transformers: TransformerInfo[],
  output: string[],
  outDir: string,
): string {
  const requireString = createRequireString(transformers);
  const configString = createConfigString(input, output, outDir);
  const transformerCalls = createTransformerCalls(transformers);

  return `${requireString}\n\nmodule.exports = ${configString.replace(
    `'<TRANSFORMERS>'`,
    transformerCalls,
  )}\n`;
}

function isValidCombination(transformers: TransformerInfo[]): boolean {
  let isValid = true;

  for (let i = 0; i < transformers.length - 1; i += 1) {
    const current = transformers[i];
    const next = transformers[i + 1];

    if (!current.to.includes(next.from)) {
      isValid = false;
      break;
    }
  }

  return isValid;
}

export async function promptForConfig(): Promise<PromptedConfig> {
  const availableTransformers = getAvailableTransformers();

  const { input } = await inquirer.prompt([
    {
      message: 'Which files do you want to transform?',
      name: 'input',
      suffix: chalk.dim(' e.g ./**/*.js'),
      type: 'input',
    },
  ]);

  const { selectedTransformers } = await inquirer.prompt([
    {
      choices: availableTransformers.map(t => ({
        name: `${t.name}\t${chalk.dim(`${t.from} -> ${t.to.join(' | ')}`)}`,
        short: t.name,
        value: t.name,
      })),
      default: [],
      message: 'Which transformers do you want to use?',
      name: 'selectedTransformers',
      type: 'checkbox',
    },
  ]);

  const transformers = availableTransformers.filter(t =>
    selectedTransformers.includes(t.name),
  );

  if (!isValidCombination(transformers)) {
    const combination = transformers.reduce(
      (str: string, transformer: TransformerInfo, i: number) => {
        if (i === 0) {
          return `${transformer.name}`;
        }

        return `${str} -> ${transformer.name}`;
      },
      '',
    );
    throw new Error(
      `Transformations in the order ${combination} is not possible`,
    );
  }

  const { output } = await inquirer.prompt([
    {
      choices: getAvailableOutputFormats(transformers),
      message: 'Which formats to you want to output?',
      name: 'output',
      type: 'checkbox',
      when: transformers.length > 0,
    },
  ]);

  const { outDir } = await inquirer.prompt([
    {
      message: 'To which directory do you want to output files?',
      name: 'outDir',
      suffix: chalk.dim(' e.g ./dist'),
      type: 'input',
    },
  ]);

  const { install } = await inquirer.prompt([
    {
      default: true,
      message: 'Do you want to install the selected transformers and the CLI?',
      name: 'install',
      type: 'confirm',
      when: transformers.length > 0,
    },
  ]);

  const { pkgManager } = await inquirer.prompt([
    {
      choices: ['npm', 'yarn'],
      message: 'Which package manager do you want to use for installation?',
      name: 'pkgManager',
      type: 'list',
      when: install === true,
    },
  ]);

  return {
    install,
    module: createModuleString(input, transformers, output, outDir),
    pkgManager,
    transformers,
  };
}
