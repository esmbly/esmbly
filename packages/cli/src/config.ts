import path from 'path';
import stringify from 'stringify-object';
import { Config, InitOptions } from '@esmbly/types';
import {
  exists,
  getRelativePathTo,
  getRoot,
  readFile,
  writeFile,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';

export const DEFAULT_FILE = '.esmblyrc.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(__dirname, `../defaults/${DEFAULT_FILE}`);
  const template = await readFile(templatePath);
  return template.toString();
}

export function getDefaultConfigPath(): string {
  const root = getRoot();
  return path.join(root, DEFAULT_FILE);
}

export async function readConfig(customPath?: string): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = require(configPath) // eslint-disable-line
  const toArray = (maybeArrayConfig: Config | Config[]): Config[] => {
    if (Array.isArray(maybeArrayConfig)) {
      return maybeArrayConfig;
    }
    return [maybeArrayConfig];
  };
  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArray(configFnResult);
  }
  return toArray(config);
}

export async function createConfig(
  options: InitOptions,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = getRoot();
  const defaultConfigPath = getDefaultConfigPath();
  if ((await exists(defaultConfigPath)) && !options.force) {
    const relativeConfigPath = getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (options.default) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template, { overwrite: options.force });
  } else {
    const config = await promptForConfig();
    const content = `module.exports = ${stringify(config)}\n`;
    await writeFile(defaultConfigPath, content, { overwrite: options.force });
  }
  return { fileName: DEFAULT_FILE, root };
}
