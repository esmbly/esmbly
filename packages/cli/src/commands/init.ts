import { Arguments, Argv } from 'yargs';
import { InitOptions } from '@esmbly/types';
import ora from 'ora';
import { createConfig } from '../utils/config';

export const command = 'init';

export const describe = 'Create an esmbly configuration file';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function builder(yargs: Argv): Argv<any> {
  return yargs
    .version(false)
    .option('default', {
      default: false,
      describe: 'Use default configuration',
      type: 'boolean',
    })
    .option('force', {
      default: false,
      describe: 'Overwrite any existing config file',
      type: 'boolean',
    })
    .help() as Argv;
}

export async function handler(argv: Arguments & InitOptions): Promise<void> {
  try {
    const { fileName, root } = await createConfig(argv);
    ora(`${fileName} created in ${root}\n`).succeed();
  } catch (err) {
    ora(`${err.message}` || err).fail();
  }
}
