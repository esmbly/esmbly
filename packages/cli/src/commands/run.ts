import { Arguments, Argv } from 'yargs';
import esmbly from '@esmbly/core';
import { Config, Transformer } from '@esmbly/types';
import { outputFactory, readFiles, transformerFactory } from '@esmbly/utils';
import stringify from 'stringify-object';
import output from '@esmbly/output';
import { readConfig } from '../config';

export interface RunOptions {
  config: string;
  silent: boolean;
  input: string[];
  transformers: string[];
  output: string[];
}

export const command = 'run';

export const describe = 'Run esmbly';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const builder = (yargs: Argv): Argv<any> => {
  return yargs
    .version(false)
    .option('config', {
      alias: 'c',
      describe: 'Configuration file path',
      type: 'string',
    })
    .option('silent', {
      default: false,
      describe: 'Silence output to the console',
      type: 'boolean',
    })
    .option('input', {
      alias: 'i',
      describe: 'The files you want to transform',
      type: 'array',
    })
    .option('transformers', {
      alias: 't',
      describe: 'The transformer you want to use',
      type: 'array',
    })
    .option('output', {
      alias: 'o',
      describe: 'The output formats you want to use',
      type: 'array',
    })
    .option('dry-run', {
      describe: 'Run transformations without outputting any files',
      type: 'boolean',
    })
    .option('print-config', {
      describe: 'Print configuration and then abort',
      type: 'boolean',
    })
    .help();
};

export const handler = async (argv: Arguments & RunOptions): Promise<void> => {
  try {
    if (argv.silent) {
      output.setErrorStream(null);
      output.setOutStream(null);
    }

    const config = await readConfig(argv.config);

    if (argv.printConfig) {
      output.out(`${stringify(config, { indent: ' ' })}\n`);
      return;
    }

    const results = await Promise.all(
      config.map(async (c: Config) => {
        const input = await readFiles(argv.input || c.input);
        const transformers = (argv.transformers || c.transformers).map(
          (transformer: string | Transformer) =>
            transformerFactory(transformer),
        );
        if (input.length < 1) {
          throw new Error(`Found 0 files matching pattern: ${c.input}`);
        }
        const output = (argv.output || c.output).map(outputFactory);
        return esmbly.run({ input, output, transformers });
      }),
    );

    if (argv.dryRun) {
      output.out(`${stringify(results, { indent: ' ' })}\n`);
    } else {
      // TODO: write files here
    }
  } catch (err) {
    output.out(`${err.message}\n` || err);
  }
};
