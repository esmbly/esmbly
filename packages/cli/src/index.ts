import yargs, { Arguments } from 'yargs';
import * as init from './commands/init';
import * as run from './commands/run';

export function runCLI(): Arguments {
  return yargs
    .scriptName('esmbly')
    .command(init)
    .command(run)
    .help().argv;
}
