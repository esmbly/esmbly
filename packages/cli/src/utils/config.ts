import path from 'path';
import { Config, InitOptions } from '@esmbly/types';
import {
  exists,
  getRelativePathTo,
  getRoot,
  readFile,
  writeFile,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';
import { installCLI, installTransformers } from './install';

export const DEFAULT_FILE = 'esmbly.config.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(
    __dirname,
    `../../defaults/${DEFAULT_FILE}`,
  );
  const template = await readFile(templatePath);
  return template.toString();
}

export function getDefaultConfigPath(): string {
  const root = getRoot();
  return path.join(root, DEFAULT_FILE);
}

type ConfigFnA = () => Config;
type ConfigFnB = () => Config[];
type ConfigFn = ConfigFnA | ConfigFnB;

function toArrayConfig(maybeArrayConfig: Config | Config[]): Config[] {
  if (Array.isArray(maybeArrayConfig)) {
    return maybeArrayConfig;
  }

  return [maybeArrayConfig];
}

export async function readConfig(
  customPath?: string,
  requirer: (requirePath: string) => unknown = require,
): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = requirer(configPath) as Config | Config[] | ConfigFn;

  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArrayConfig(configFnResult);
  }

  return toArrayConfig(config);
}

export async function createConfig(
  options: InitOptions,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = getRoot();
  const defaultConfigPath = getDefaultConfigPath();

  if ((await exists(defaultConfigPath)) && !options.force) {
    const relativeConfigPath = getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (options.default) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template, { overwrite: options.force });
  } else {
    const config = await promptForConfig();

    if (config.install) {
      await installTransformers(config.transformers, config.pkgManager);
      await installCLI(config.pkgManager);
    }

    await writeFile(defaultConfigPath, config.module, {
      overwrite: options.force,
    });
  }

  return { fileName: DEFAULT_FILE, root };
}
