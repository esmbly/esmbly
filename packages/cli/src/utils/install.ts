import { exec } from 'child_process';
import { promisify } from 'util';
import { TransformerInfo } from '@esmbly/utils';
import ora from 'ora';

export function getPkgManagerCommand(pkgManager: string): string {
  switch (pkgManager) {
    case 'npm':
      return 'npm install --save-dev';
    case 'yarn':
      return 'yarn add --dev';
    default:
      throw new Error(`Package manager ${pkgManager} is not supported`);
  }
}

export async function installTransformers(
  transformers: TransformerInfo[],
  pkgManager: string,
): Promise<void> {
  const command = getPkgManagerCommand(pkgManager);

  for (const transformer of transformers) {
    const spinner = ora(`Installing: ${transformer.pkg}`).start();
    await promisify(exec)(`${command} ${transformer.pkg}`);
    spinner.succeed(`Installed: ${transformer.pkg}`);
  }
}

export async function installCLI(pkgManager: string): Promise<void> {
  const command = getPkgManagerCommand(pkgManager);

  const spinner = ora('Installing: @esmbly/cli').start();
  await promisify(exec)(`${command} @esmbly/cli`);
  spinner.succeed('Installed: @esmbly/core');
}
