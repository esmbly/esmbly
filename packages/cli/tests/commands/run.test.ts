import esmbly from '@esmbly/core';
import yargs from 'yargs';
import * as utils from '@esmbly/utils';
import { Format } from '@esmbly/types';
import * as run from '../../src/commands/run';
import * as config from '../../src/config';
import CommandRunner from '../__fixtures__/CommandRunner';
import MockTransformer from '../__fixtures__/FooTransformer';
import MockFiles from '../__fixtures__/files';
import MockConfig from '../__fixtures__/config';

jest.mock('@esmbly/printer');

const command = new CommandRunner(run);

const setup = (): {
  readConfigSpy: jest.SpyInstance;
  readFilesSpy: jest.SpyInstance;
  transformerFactorySpy: jest.SpyInstance;
  tearDown: () => void;
} => {
  const readConfigSpy = jest.spyOn(config, 'readConfig');
  readConfigSpy.mockResolvedValue(MockConfig);
  const readFilesSpy = jest.spyOn(utils, 'readFiles');
  readFilesSpy.mockResolvedValue(MockFiles);
  const transformerFactorySpy = jest.spyOn(utils, 'transformerFactory');
  transformerFactorySpy.mockReturnValue(MockTransformer());
  return {
    readConfigSpy,
    readFilesSpy,
    tearDown: () => {
      readConfigSpy.mockRestore();
      readFilesSpy.mockRestore();
      transformerFactorySpy.mockRestore();
    },
    transformerFactorySpy,
  };
};

describe('Run', () => {
  it('exposes the correct command', () => {
    expect(run.command).toEqual('run');
  });

  it('exposes the correct description', () => {
    expect(run.describe).toEqual('Run esmbly');
  });

  it('exposes the correct help information', async () => {
    yargs.scriptName('esmbly');
    const output = await command.run('run --help');
    expect(output).toMatchSnapshot();
  });

  it('reads the config', async () => {
    const { readConfigSpy, tearDown } = setup();
    await command.run('run');
    expect(readConfigSpy).toHaveBeenCalledTimes(1);
    tearDown();
  });

  it('reads all files', async () => {
    const { readFilesSpy, tearDown } = setup();
    await command.run('run');
    // TODO: remove this workaround (issue: https://github.com/yargs/yargs/issues/1069)
    await new Promise(resolve => setTimeout(resolve, 100));
    expect(readFilesSpy).toHaveBeenCalledTimes(1);
    expect(readFilesSpy).toHaveBeenCalledWith(['**/*.js']);
    tearDown();
  });

  it('requires all transformers', async () => {
    const { transformerFactorySpy, tearDown } = setup();
    await command.run('run');
    // TODO: remove this workaround (issue: https://github.com/yargs/yargs/issues/1069)
    await new Promise(resolve => setTimeout(resolve, 100));
    expect(transformerFactorySpy).toHaveBeenCalledTimes(1);
    tearDown();
  });

  it('runs esmbly', async () => {
    const { tearDown } = setup();
    const runSpy = jest.spyOn(esmbly, 'run');
    await command.run('run');
    // TODO: remove this workaround (issue: https://github.com/yargs/yargs/issues/1069)
    await new Promise(resolve => setTimeout(resolve, 100));
    expect(runSpy).toHaveBeenCalledTimes(1);
    expect(runSpy).toHaveBeenCalledWith({
      input: MockFiles,
      output: [{ format: Format.WebAssembly }],
      transformers: [
        {
          inputFormat: 'Any',
          outputFormats: ['Flow'],
          transform: expect.any(Function),
        },
      ],
    });
    runSpy.mockRestore();
    tearDown();
  });

  it.todo('add tests for cli flags');

  it.todo('test output by mocking @esmbly/output');

  it.todo('test error output by mocking @esmbly/output');
});
