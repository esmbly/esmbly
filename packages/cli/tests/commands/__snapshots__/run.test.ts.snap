// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run exposes the correct help information 1`] = `
"esmbly run

Run Esmbly

Options:
  --config, -c        Configuration file path                           [string]
  --silent            Silence output to the console   [boolean] [default: false]
  --input, -i         The files you want to transform                    [array]
  --transformers, -t  The transformer you want to use                    [array]
  --output, -o        The output formats you want to use                 [array]
  --dry-run           Run transformations without outputting any files [boolean]
  --print-config      Print configuration and then abort               [boolean]
  --help              Show help                                        [boolean]"
`;
